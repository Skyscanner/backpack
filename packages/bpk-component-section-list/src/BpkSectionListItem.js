/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import PropTypes from 'prop-types';
import React, { type Node } from 'react';

import { cssModules } from '../../bpk-react-utils';
import BpkLargeChevronRightIcon from '../../bpk-component-icon/lg/chevron-right';
import { withRtlSupport } from '../../bpk-component-icon';

import STYLES from './BpkSectionListItem.module.scss';

const BpkLargeChevronRightIconWithRtlSupport = withRtlSupport(
  BpkLargeChevronRightIcon,
);

const getClassName = cssModules(STYLES);

type Props = {
  blank: boolean,
  children: Node,
  className: ?string,
  href: ?string,
  onClick: ?(event: SyntheticEvent<>) => void,
};

const BpkSectionListItem = (props: Props) => {
  const { blank, children, className, href, onClick, ...rest } = props;
  const classNames = [
    getClassName(
      'bpk-section-list-item',
      (href || onClick) && 'bpk-section-list-item--interactive',
      className,
    ),
  ];

  if (href) {
    const target = blank ? '_blank' : null;
    return (
      // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
      <a
        href={href}
        target={target}
        onClick={onClick}
        className={classNames.join(' ')}
        {...rest}
      >
        {children}
        <BpkLargeChevronRightIconWithRtlSupport
          className={getClassName('bpk-section-list-item__chevron')}
        />
      </a>
    );
  }

  if (onClick) {
    return (
      // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
      <button
        type="button"
        onClick={onClick}
        className={classNames.join(' ')}
        {...rest}
      >
        {children}
        <BpkLargeChevronRightIconWithRtlSupport
          className={getClassName('bpk-section-list-item__chevron')}
        />
      </button>
    );
  }

  return (
    // $FlowFixMe[cannot-spread-inexact] - inexact rest. See 'decisions/flowfixme.md'.
    <div className={classNames.join(' ')} {...rest}>
      {children}
    </div>
  );
};

BpkSectionListItem.propTypes = {
  children: PropTypes.node.isRequired,
  blank: PropTypes.bool,
  className: PropTypes.string,
  href: PropTypes.string,
  onClick: PropTypes.func,
};

BpkSectionListItem.defaultProps = {
  blank: false,
  className: null,
  href: null,
  onClick: null,
};

export default BpkSectionListItem;
