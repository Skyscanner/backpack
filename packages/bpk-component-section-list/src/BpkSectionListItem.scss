/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2019 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

@import '~bpk-mixins/index';

.bpk-section-list-item {
  $root: &;

  display: flex;
  width: 100%;
  min-height: $bpk-spacing-lg * 2;
  padding: $bpk-spacing-xs $bpk-spacing-sm;
  justify-content: space-between;
  align-items: center;
  border: none;
  background-color: $bpk-color-white;
  color: $bpk-color-gray-700;
  text-decoration: none;

  @include bpk-apply-gray-700-to(color);
  @include bpk-border-bottom-sm($bpk-color-gray-100);
  @include bpk-border-bottom-sm(var(--bpk-color-gray-100, $bpk-color-gray-100));

  &--interactive {
    cursor: pointer;
  }

  &__chevron {
    @include bpk-apply-gray-500-to(fill);

    #{$root}:hover & {
      @include bpk-apply-gray-700-to(fill);
    }

    #{$root}:active & {
      @include bpk-apply-gray-900-to(fill);
    }
  }
}
