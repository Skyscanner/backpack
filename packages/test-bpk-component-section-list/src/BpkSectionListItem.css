@keyframes bpk-keyframe-spin{100%{transform:rotate(1turn)}}.bpk-section-list-item{display:flex;width:100%;min-height:3.75rem;padding:.375rem .75rem;justify-content:space-between;align-items:center;border:none;background-color:#fff;color:#444560;text-decoration:none;box-shadow:0 -1px 0 0 #dddde5 inset}.bpk-section-list-item--interactive{cursor:pointer}.bpk-section-list-item__chevron{fill:#68697f}.bpk-section-list-item:hover .bpk-section-list-item__chevron{fill:#444560}.bpk-section-list-item:active .bpk-section-list-item__chevron{fill:#111236}
