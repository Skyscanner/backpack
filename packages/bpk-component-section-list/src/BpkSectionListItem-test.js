/*
 * Backpack - Skyscanner's Design System
 *
 * Copyright 2016-2019 Skyscanner Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/* @flow strict */

import React from 'react';
import renderer from 'react-test-renderer';

import BpkSectionListItem from './BpkSectionListItem';

describe('BpkSectionListItem', () => {
  it('should render correctly', () => {
    const tree = renderer
      .create(<BpkSectionListItem>Hello world</BpkSectionListItem>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a "href" prop', () => {
    const tree = renderer
      .create(<BpkSectionListItem href="#">Hello world</BpkSectionListItem>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with an "onClick" prop', () => {
    const tree = renderer
      .create(
        <BpkSectionListItem onClick={jest.fn()}>
          Hello world
        </BpkSectionListItem>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with a custom className', () => {
    const tree = renderer
      .create(
        <BpkSectionListItem className="custom-class">
          Hello world
        </BpkSectionListItem>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render correctly with arbitrary props', () => {
    const tree = renderer
      .create(<BpkSectionListItem testId="123">Hello world</BpkSectionListItem>)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
