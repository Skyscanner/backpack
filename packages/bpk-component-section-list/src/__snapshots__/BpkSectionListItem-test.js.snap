// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BpkSectionListItem should render correctly 1`] = `
<div
  className="bpk-section-list-item"
>
  Hello world
</div>
`;

exports[`BpkSectionListItem should render correctly with a "href" prop 1`] = `
<a
  className="bpk-section-list-item bpk-section-list-item--interactive"
  href="#"
  onClick={null}
  target={null}
>
  Hello world
  <svg
    className="bpk-section-list-item__chevron bpk-icon--rtl-support"
    height="24"
    style={
      Object {
        "height": "1.5rem",
        "width": "1.5rem",
      }
    }
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M8.293 4.293a1 1 0 0 0 0 1.414L14.586 12l-6.293 6.293a1 1 0 1 0 1.414 1.414l7-7a1 1 0 0 0 0-1.414l-7-7a1 1 0 0 0-1.414 0z"
      fillRule="evenodd"
    />
  </svg>
</a>
`;

exports[`BpkSectionListItem should render correctly with a custom className 1`] = `
<div
  className="bpk-section-list-item custom-class"
>
  Hello world
</div>
`;

exports[`BpkSectionListItem should render correctly with an "onClick" prop 1`] = `
<button
  className="bpk-section-list-item bpk-section-list-item--interactive"
  onClick={[MockFunction]}
  type="button"
>
  Hello world
  <svg
    className="bpk-section-list-item__chevron bpk-icon--rtl-support"
    height="24"
    style={
      Object {
        "height": "1.5rem",
        "width": "1.5rem",
      }
    }
    width="24"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M8.293 4.293a1 1 0 0 0 0 1.414L14.586 12l-6.293 6.293a1 1 0 1 0 1.414 1.414l7-7a1 1 0 0 0 0-1.414l-7-7a1 1 0 0 0-1.414 0z"
      fillRule="evenodd"
    />
  </svg>
</button>
`;

exports[`BpkSectionListItem should render correctly with arbitrary props 1`] = `
<div
  className="bpk-section-list-item"
  testId="123"
>
  Hello world
</div>
`;
